Migrate from Terraform to OpenTofu

davidtabat opened the pull request:
As part of this PR, I renamed the terraform repository to opentofu in variables.tf for consistency.

Please ensure the repository on GitHub is renamed from terraform to opentofu before merging this PR. Otherwise, merging will trigger the creation of a new repository named opentofu.

If you don't want renaming the repository, let me know, and I will revert the name in variables.tf.

sprankhub reviewed (commented):
First, thanks a lot for your PR!

This looks too simple to be true :joy: Did you test this somehow, by chance? Not sure, but who knows - maybe there is a way :)

So you think renaming this repository and then merging this PR should be enough? Or are there additional steps we need to take?

sprankhub commented:
/plan

sprankhub commented:
@davidtabat the "plan" above shows an error, even though it looks okay. Any idea? Do you think it will be gone when we rename this repository?

fballiano commented:
genuine question: if we rename the repo, shouldn't we rename it to a platform-agnostic one like "infrastructure" or something like that?

davidtabat commented:
@sprankhub Yes I tested it. Described in the taiga ticket comment - how exactly https://tree.taiga.io/project/mage-os-mage-os-technical-initiatives/us/37?kanban-status=9349607

davidtabat commented:
@fballiano Yes, agree with you, that indeed sounds better.

sprankhub commented:
I agree, but https://github.com/mage-os/infrastructure is taken already :D Other ideas?

* `infrastructure-setup`
* `github-setup`
* `permissions`
* `github-permissions`
* ...

davidtabat commented:
I tested renaming on a dummy repo, and even after renaming it on GitHub, there were still errors due to the state saved in the backend. For now, I’ll take some time to figure out how to do it properly. 😄

Regarding the name, I think 'infrastructure-setup' is good. Maybe we should create a poll in Discord? Or is this too small of a matter to give so much attention to?"

DavidLambauer commented:
Are we sure that this switch won't result in a lock change or even worse in a re-creation of existing repos?

DavidLambauer commented:
So, for example, when we destroy the tf repo, can this have an impact on any other repos?

davidtabat commented:
@DavidLambauer It shouldn't, but I will test that here to be sure ofc: https://github.com/orgs/test-opentofu-mageos/repositories.

fballiano commented:
> Regarding the name, I think 'infrastructure-setup' is good. 

I'm nobody to decide anything but I'm fine with infrastructure-setup

> Are we sure that this switch won't result in a lock change or even worse in a re-creation of existing repos?

I sincerely hope TF doesn't work that way, is there a way to know?

DavidLambauer commented:
In the early days, I ran into some lock issues, and that's why I am asking. If it's properly tested, we should be fine I guess.

fballiano commented:
Hi guys, as we were saying the the last tech meeting, can we leave the terraform repository name as it is? could we merge this PR as in it in that case?

sprankhub commented:
I guess it should be fine then, yes. @davidtabat, could you update the PR accordingly?

davidtabat commented:
Already on it, there were also some other PRs merged while this one was open, so I had to pull updates and fix merge conflicts. After that I started re-testing and I have some error messages on plan and apply workflows. I will have to look into them.

davidtabat commented:
@sprankhub done, before merging this organization can be used for testing: https://github.com/mageos-opentofu-test/terraform

sprankhub commented on main.tf:
```diff
@@ -176,15 +176,15 @@ resource "github_team_repository" "tech-lead" {
 
 resource "github_repository_file" "codeowners" {
-  for_each   = var.repositories
+  for_each   = { for k, v in var.repositories : k => v if lookup(v, "archived", false) == false }
```
Minor: This could be unified with https://github.com/mage-os/terraform/blob/main/main.tf#L240-L243 I guess? Not sure which one is better, I just noticed that there is a difference :)

davidtabat replied:
My approach iterates through the repositories defined in the variables file. If the "archived" property is explicitly set to true, the repository is considered archived. If the property is missing, it defaults to false. The approach you linked also iterates through var.repositories, but it then tries to reference the actual GitHub resource, which may not exist yet. I assume this would cause a failure in the plan workflow if a new repository is added to the variables file but hasn't been created on GitHub yet. Therefore, I recommend using my approach. Let me know if I’m mistaken.

sprankhub commented:
/plan

sprankhub commented:
Okay so let's have a look at the changes...

1. `dashboard`: `vulnerability_alerts` are set from `true` to `null`... That looks weird. But maybe this would even happen with Terraform. Maybe someone set this property manually to `true` on the repository? I guess it would be okay to ignore this for now.
2. `mage-os-website`, `magento-gitpod`: These will be destroyed. According to https://github.com/mage-os/terraform/blob/main/main.tf#L109-L112, I think this shouldn't happen. My guess was that OpenTofu doesn't support this, but it's [part of the docs](https://opentofu.org/docs/language/meta-arguments/lifecycle/#syntax-and-arguments). Any idea?
3. Change of commit messages in all repositories: All fine.

I think 2 is the only critical issue for now. 1 and 3 seem to be fine to me.

fballiano commented:
@davidtabat sorry to bother, do you think we could solve what @sprankhub reported?

davidtabat commented:
@fballiano Sure, I will do that this week.

davidtabat commented:
@sprankhub 
As I understand it, the message doesn’t indicate that the mage-os-website and magento-gitpod repositories will be destroyed. Instead, it means that the codeowner resource for those repositories will be removed from the backend.

This happens because archived repositories are skipped in the for_each loop of the codeowners block in my PR. I added this condition to skip archived repos after encountering the "Repository was archived so is read-only" error. The issue arose when I changed the commit message for the CODEOWNERS file from "Managed by Terraform" to "Managed by OpenTofu." OpenTofu tried to apply this change, but since the repo was archived, it threw the error. I resolved this by skipping archived repositories in the codeowners resource block.

If you prefer, I can revert the commit message back to "Managed by Terraform" so that OpenTofu won’t try to modify archived repositories. In that case, there would be no need for this condition, and the codeowner resource won’t be destroyed either.

rhoerr commented:
> If you prefer, I can revert the commit message back to "Managed by Terraform" so that OpenTofu won’t try to modify archived repositories. In that case, there would be no need for this condition, and the codeowner resource won’t be destroyed either.

Personally, that sounds better to me. That would also be consistent with keeping the repository name `terraform`, and will be more easily recognizable.

sprankhub commented:
Thanks for your feedback, @davidtabat! You are indeed right, I misread the log. Sorry for this. Let's indeed keep using "Managed by Terraform", so that we don't have unnecessary updates on all repositories. I'll have a look at another plan output then, but I guess we are good to go then!

By the way, I think keeping the condition for archived repositories for the `CODEOWNERS` file does make sense, so I personally think we should keep it.

davidtabat commented:
@sprankhub If we keep that condition for archived repos - the codeowner resource will still be destroyed, is that ok?

In order to not destroy it, we need to remove this condition I added, and also revert commit message back. Only reverting commit message, won't prevent the codeowner resource from being destroyed for archived repos.

sprankhub commented:
That's a tricky one. If we keep the condition, OpenTofu will attempt to remove the file, which will fail, because the repo is archived. We would need to fix this manually, then. Not sure what is better - feel free to decide and go for it :)

davidtabat commented:
I don't have the necessary permissions to make manual changes on the Mage-OS GitHub repository.  

Additionally, I believe it's best to keep this PR focused solely on the Terraform-to-OpenTofu migration, rather than introducing general code improvements. Mixing logical changes with the migration could make it harder to identify the root cause of any potential bugs.

Therefore I will just revert both of my changes: commit message and for loop modification as well.

davidtabat commented:
Done, let's try running plan workflow again.

sprankhub commented:
/plan

rhoerr commented:
The changes look expected to me, with the caveat that I don't know what the hash changes mean in https://github.com/mage-os/terraform/pull/104/files#diff-ad2c33b49112c808ea24699b01e9e020735e894c9438ace109477b3e6098cc43 .

That said, I have no direct experience with Terraform beyond PRs to this same repo.

Thank you David.

fballiano commented:
@sprankhub @Vinai what should we do here? I mean, we came this far, it makes sense to finish it right? otherwise it would have been better to not even start :-)

Vinai commented:
I trust @sprankhub and @davidtabat things are okay, even though I also have no idea what the hashes mean.
YOLO!

rhoerr commented:
I read into `.terraform.lock.hcl` a bit, it's a dependency lock file like `composer.lock` or `packages.lock` for other systems. It follows that the package hashes would change going from Terraform to OpenTofu.
